In [125]:
#Import data using pydataset
from pydataset import data as d
import math

In [82]:
d('mpg', show_doc=True) # view the documentation for the dataset
mpg = d('mpg') # load the dataset and store it in a variable

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




#  <center>1</center></h1>
<center>Copy the code from the lesson to create a dataframe full of student grades.</center></h1>

In [83]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(df)

pandas.core.frame.DataFrame

In [84]:
df

name  math  english  reading
0     Sally    62       85       80
1      Jane    88       79       67
2     Suzie    94       74       95
3     Billy    98       96       88
4       Ada    77       92       98
5      John    79       76       93
6    Thomas    82       64       81
7     Marie    93       63       90
8    Albert    92       62       87
9   Richard    69       80       94
10    Isaac    92       99       93
11     Alan    92       62       72

In [85]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   name     12 non-null     object
 1   math     12 non-null     int64 
 2   english  12 non-null     int64 
 3   reading  12 non-null     int64 
dtypes: int64(3), object(1)
memory usage: 512.0+ bytes


In [86]:
# 1.a Create a column named passing_english that indicates whether each student has a passing grade in english.
df['passing_english'] = df.english >= 70
df

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72            False

In [87]:
# 1.b Sort the english grades by the passing_english column. 
'''How are duplicates handled? 
Duplicates are handled by alphabetical order of the first column of the record'''

df = df.sort_values(by='passing_english')
df

name  math  english  reading  passing_english
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True

In [88]:
# 1.c Sort the english grades first by passing_english and then by student name. All the students that are failing english should be first, and within the students that are failing english they should be ordered alphabetically. The same should be true for the students passing english. (Hint: you can pass a list to the .sort_values method)
df1c = df.sort_values(by=['passing_english', 'name'])
df1c

name  math  english  reading  passing_english
11     Alan    92       62       72            False
8    Albert    92       62       87            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
1      Jane    88       79       67             True
5      John    79       76       93             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
2     Suzie    94       74       95             True

In [89]:
# 1.d Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.
df1d = df.sort_values(by=['passing_english', 'english'], ascending=[False, False])
df1d

name  math  english  reading  passing_english
10    Isaac    92       99       93             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
0     Sally    62       85       80             True
9   Richard    69       80       94             True
1      Jane    88       79       67             True
5      John    79       76       93             True
2     Suzie    94       74       95             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False

In [90]:
# 1.e Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.
df['OVERALL'] = ((df.math + df.english + df.reading) / 3).round(decimals=2)
df['Passed'] = df.OVERALL > 70
df

name  math  english  reading  passing_english  OVERALL  Passed
6    Thomas    82       64       81            False    75.67    True
7     Marie    93       63       90            False    82.00    True
8    Albert    92       62       87            False    80.33    True
11     Alan    92       62       72            False    75.33    True
0     Sally    62       85       80             True    75.67    True
1      Jane    88       79       67             True    78.00    True
2     Suzie    94       74       95             True    87.67    True
3     Billy    98       96       88             True    94.00    True
4       Ada    77       92       98             True    89.00    True
5      John    79       76       93             True    82.67    True
9   Richard    69       80       94             True    81.00    True
10    Isaac    92       99       93             True    94.67    True

## 

# <center>2</center></h1>

<center>Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:</center></h1>

#### How many rows and columns are there?
There are 234 rows and 11 colums

In [91]:
mpg.shape

(234, 11)

#### What are the data types of each column? 

In [92]:
mpg.dtypes

manufacturer     object
model            object
displ           float64
year              int64
cyl               int64
trans            object
drv              object
cty               int64
hwy               int64
fl               object
class            object
dtype: object

#### Summarize the dataframe with .info and .describe

In [93]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


In [94]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

In [95]:
# Rename the cty column to city.
mpg = mpg.rename(columns={'cty':'city'})
mpg

manufacturer   model  displ  year  cyl       trans drv  city  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16   26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [96]:
# Rename the hwy column to highway.
mpg = mpg.rename(columns={'hwy':'highway'})
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [97]:
# Do any cars have better city mileage than highway mileage?
# NO

mpg[mpg.city > mpg.highway]
   

Empty DataFrame
Columns: [manufacturer, model, displ, year, cyl, trans, drv, city, highway, fl, class]
Index: []

In [98]:
mpg_higher_city = mpg[mpg.city > mpg.highway]

In [99]:
mpg_higher_city

Empty DataFrame
Columns: [manufacturer, model, displ, year, cyl, trans, drv, city, highway, fl, class]
Index: []

In [100]:
# Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.
mpg['milage_difference'] = mpg.highway - mpg.city
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  milage_difference  
1    compact                 11  
2    compact                  8  
3    compact                 11  
4    compact                  9  
5    compact                 10  
..       ...                ...  
230  midsize                  9  
231  midsize                  8  
232  midsize                 10  
233  midsize                  8  
234  midsize                  9  

[234 rows x 12 columns]

In [102]:
# Which car (or cars) has the highest mileage difference?
mpg[mpg.milage_difference == mpg.milage_difference.max()]

manufacturer       model  displ  year  cyl     trans drv  city  highway  \
107        honda       civic    1.8  2008    4  auto(l5)   f    24       36   
223   volkswagen  new beetle    1.9  1999    4  auto(l4)   f    29       41   

    fl       class  milage_difference  
107  c  subcompact                 12  
223  d  subcompact                 12

In [120]:
# Which compact class car has the lowest highway mileage? The best?
lowest_compact_mpg = mpg[mpg['class'] == 'compact']
lowest_compact_mpg[lowest_compact_mpg['milage_difference'] == lowest_compact_mpg['milage_difference'].max()]


manufacturer  model  displ  year  cyl       trans drv  city  highway fl  \
1           audi     a4    1.8  1999    4    auto(l5)   f    18       29  p   
3           audi     a4    2.0  2008    4  manual(m6)   f    20       31  p   
213   volkswagen  jetta    1.9  1999    4  manual(m5)   f    33       44  d   

       class  milage_difference  
1    compact                 11  
3    compact                 11  
213  compact                 11

In [131]:
# Create a column named average_mileage that is the mean of the city and highway mileage.
mpg['average_mileage'] = (mpg.city + mpg.highway) / 2
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  milage_difference  average_mileage  
1    compact                 11             23.5  
2    compact                  8             25.0  
3    compact                 11             25.5  
4    compact                  9             25.5  
5    compact                 10             21.0  
..       ...                ...              ...  
230  midsize                  9             23.5  
231  midsize                  8             25.0  
232  midsize                 10             21.0  
233  midsize                  8             22.0  
234  midsize                  9             21.5  

[234 rows x 13 columns]

In [144]:
# Which dodge car has the best average mileage? 
dodge_df = mpg[mpg['manufacturer'] == 'dodge']
dodge_df[dodge_df['average_mileage'] == dodge_df['average_mileage'].max()]

manufacturer        model  displ  year  cyl     trans drv  city  highway  \
38        dodge  caravan 2wd    2.4  1999    4  auto(l3)   f    18       24   

   fl    class  milage_difference  average_mileage  
38  r  minivan                  6             21.0

In [145]:
#The worst?
dodge_df[dodge_df['average_mileage'] == dodge_df['average_mileage'].min()] 

manufacturer                model  displ  year  cyl       trans drv  city  \
55        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
60        dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
66        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
70        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   

    highway fl   class  milage_difference  average_mileage  
55       12  e  pickup                  3             10.5  
60       12  e     suv                  3             10.5  
66       12  e  pickup                  3             10.5  
70       12  e  pickup                  3             10.5

# <center>3</center></h1>

<center>Load the Mammals dataset. Read the documentation for it, and use the data to answer these questions:</center></h1>

In [146]:
from pydataset import data 
df = data('Mammals')

data('Mammals', show_doc=True)

Mammals

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Garland(1983) Data on Running Speed of Mammals

### Description

Observations on the maximal running speed of mammal species and their body
mass.

### Usage

    data(Mammals)

### Format

A data frame with 107 observations on the following 4 variables.

weight

Body mass in Kg for "typical adult sizes"

speed

Maximal running speed (fastest sprint velocity on record)

hoppers

logical variable indicating animals that ambulate by hopping, e.g. kangaroos

specials

logical variable indicating special animals with "lifestyles in which speed
does not figure as an important factor": Hippopotamus, raccoon (Procyon),
badger (Meles), coati (Nasua), skunk (Mephitis), man (Homo), porcupine
(Erithizon), oppossum (didelphis), and sloth (Bradypus)

### Details

Used by Chappell (1989) and Koenker, Ng and Portnoy (1994) to illustrate the
fitting of piecewise linear curves.

### Source

Garland, T. (

In [148]:
# How many rows and columns are there?
f" There are {df.shape[0]} rows and  {df.shape[1]} columns."

' There are 107 rows and  4 columns.'

In [149]:
# What are the data types?

df.dtypes

weight      float64
speed       float64
hoppers        bool
specials       bool
dtype: object

In [150]:
# Summarize the dataframe with .info and .describe
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


In [154]:
df.describe()


weight       speed
count   107.000000  107.000000
mean    278.688178   46.208411
std     839.608269   26.716778
min       0.016000    1.600000
25%       1.700000   22.500000
50%      34.000000   48.000000
75%     142.500000   65.000000
max    6000.000000  110.000000

In [160]:
# What is the the weight of the fastest animal?
df.head()
fastest = df[df['speed'] == df['speed'].max()]
fastest

weight  speed  hoppers  specials
53    55.0  110.0    False     False

In [170]:
# What is the overal percentage of specials?
df.specials.sum() / len(df) 

0.09345794392523364

In [174]:
# How many animals are hoppers that are above the median speed? What percentage is this?
hoppers = df[df['hoppers'] == True]
hoppers

weight  speed  hoppers  specials
82    0.056   21.0     True     False
85    0.035   32.0     True     False
86    0.035   14.0     True     False
96    4.600   64.0     True     False
97    4.400   72.0     True     False
98    4.000   72.0     True     False
99    3.500   56.0     True     False
100   2.000   64.0     True     False
101   1.900   56.0     True     False
102   1.500   50.0     True     False
103   1.500   40.0     True     False

In [175]:
median_speed = df['speed'].median()
median_speed

48.0

In [177]:
hoppers

weight  speed  hoppers  specials
82    0.056   21.0     True     False
85    0.035   32.0     True     False
86    0.035   14.0     True     False
96    4.600   64.0     True     False
97    4.400   72.0     True     False
98    4.000   72.0     True     False
99    3.500   56.0     True     False
100   2.000   64.0     True     False
101   1.900   56.0     True     False
102   1.500   50.0     True     False
103   1.500   40.0     True     False

In [188]:
above_the_median_speed = hoppers[hoppers['speed'] >= median_speed].shape[0]

In [191]:
f"{(above_the_median_speed / hoppers.shape[0])}%" 

'0.6363636363636364%'

In [199]:
f"{round(above_the_median_speed / df.shape[0] * 100, 2)}% are hoppers that are above the median speed"

'6.54% are hoppers that are above the median speed'